aioxmpp/bookmarks/service.py

```diff
@@ -57,29 +92,256 @@ class BookmarkClient(service.Service):
+
+        # group the bookmarks into groups whose elements may transform
+        # among one another by on_bookmark_changed events. This information
+        # is given by the type of the bookmark and the .primary property
+        changable_groups = {}
+
+        for item in self._bookmark_cache:
+            group = changable_groups.setdefault(
+                (type(item), item.primary),
+                ([], [])
+            )
+            group[0].append(item)
+
+        for item in new_bookmarks:
+            group = changable_groups.setdefault(
+                (type(item), item.primary),
+                ([], [])
             )
-        yield from self._private_xml.set_private_xml(bookmarks)
+            group[1].append(item)
+
+        for old, new in changable_groups.values():
+
+            # the first branches are fast paths which should catch
+            # most cases – especially all cases where each bare jid of
+            # a conference bookmark or each url of an url bookmark is
+            # only used in one bookmark
+            if len(old) == len(new) == 1:
+                old_entry = old.pop()
+                new_entry = new.pop()
+                if old_entry == new_entry:
+                    # the bookmark is unchanged, do not emit an event
+                    pass
+                else:
+                    self.logger.debug("changed %s -> %s", old_entry, new_entry)
+                    self.on_bookmark_changed(old_entry, new_entry)
+            elif len(new) == 0:
+                for removed in old:
+                    self.logger.debug("removed %s", removed)
+                    self.on_bookmark_removed(removed)
+            elif len(old) == 0:
+                for added in new:
+                    self.logger.debug("added %s", added)
+                    self.on_bookmark_added(added)
+            else:
+                old, new = subdivide(0, old, new)
+
+                assert len(old) == 0 or len(new) == 0
+
+                for removed in old:
+                    self.logger.debug("removed %s", removed)
+                    self.on_bookmark_removed(removed)
+
+                for added in new:
+                    self.logger.debug("added %s", added)
+                    self.on_bookmark_added(added)
+
+        self._bookmark_cache = new_bookmarks
+
+    @asyncio.coroutine
+    def get_bookmarks(self):
+        """
+        Get the stored bookmarks from the server. Causes signals to be
+        fired to reflect the changes.
+
+        :returns: a list of bookmarks
+        """
+        bookmarks = yield from self._get_bookmarks()
+        self._diff_emit_update(bookmarks)
+        return bookmarks
+
+    @asyncio.coroutine
+    def set_bookmarks(self, bookmarks):
+        """
+        Store the sequence of bookmarks `bookmarks`.
+
+        Causes signals to be fired to reflect the changes.
+
+        .. note:: This should normally not be used unless you have to
```

Comment:
Can we rephrase that so that the emphasis is on the fact that this is only safe when no other clients are modifying the bookmarks? In all other cases, the normal methods should really be used.